libs/common/RespReadUtils.cs

```diff
@@ -4,156 +4,372 @@
 using System;
 using System.Buffers;
 using System.Buffers.Text;
-using System.Diagnostics;
 using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
 using System.Text;
+using Garnet.common.Parsing;
 
 namespace Garnet.common
 {
     /// <summary>
-    /// Utilities for reading RESP protocol
+    /// Utilities for reading RESP protocol messages.
     /// </summary>
     public static unsafe class RespReadUtils
     {
         /// <summary>
-        /// Get Header length
+        /// Tries to read the leading sign of the given ASCII-encoded number.
         /// </summary>
-        /// <param name="len"></param>
-        /// <param name="ptr"></param>
-        /// <param name="end"></param>
-        /// <returns></returns>
-        public static bool ReadHeaderLength(out int len, ref byte* ptr, byte* end)
+        /// <param name="ptr">String to try reading sign from.</param>
+        /// <param name="negative">Whether the sign is '-'.</param>
+        /// <returns>True if either '+' or '-' was found, false otherwise.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool TryReadSign(byte* ptr, out bool negative)
         {
-            len = -1;
-            if (ptr + 3 >= end)
-                return false;
+            negative = (*ptr == '-');
+            return negative || (*ptr == '+');
+        }
 
-            Debug.Assert(*ptr == '$');
-            ptr++;
-            bool neg = *ptr == '-';
-            int ksize = *ptr++ - '0';
-            while (*ptr != '\r')
+        /// <summary>
+        /// Tries to read an unsigned 64-bit integer from a given ASCII-encoded input stream.
+        /// </summary>
+        /// <param name="ptr">Pointer to the beginning of the ASCII encoded input string.</param>
+        /// <param name="end">The end of the string to parse.</param>
+        /// <param name="value">If parsing was successful, contains the parsed ulong value.</param>
+        /// <param name="bytesRead">If parsing was successful, contains the number of bytes that were parsed.</param>
+        /// <returns>
+        /// True if a ulong was successfully parsed, false if the input string did not start with
+        /// a valid integer or the end of the string was reached before finishing parsing.
+        /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool TryReadUlong(ref byte* ptr, byte* end, out ulong value, out ulong bytesRead)
+        {
+            bytesRead = 0;
+            value = 0;
+            var readHead = ptr;
+
+            // Fast path for the first 19 digits.
+            // NOTE: UINT64 overflows can only happen on digit 20 or later (if integer contains leading zeros).
```

Comment:
I don't think it'll simplify the code significantly unless there is a way to check for the overflow after it has occurred. If you have any ideas for this, please let me know.

The reason I split the loop is to reduce the number of checks for integer overflows (or checks for whether we need to check). But if we have a 20-digit number we will always have to check for an overflow on the last digit (NOTE: We do not know how long the number is until we are done parsing - so we cannot use that for an initial shortcut).

Reply:
I get your point, but I was thinking that since we expect to read a ulong, we should eliminate the second loop because we can have at most 20 digits/bytes.
First read up to 19 digits/bytes (as it is done now), then check 20th digit/byte for overflow. If 21st byte is stil a digit then fail.
I might be missing something here, then semantics of this method are slightly confusing.

Reply:
See [below](https://github.com/microsoft/garnet/pull/332#discussion_r1583923524). Using a loop for the last digit allows us to support leading zeros for robustness and doesn't come at a visible performance cost.